### PR TITLE
Add encryption conform EIP1024

### PR DESCRIPTION
**Pull request by ronaldmannak**
This is a draft PR

Encryption is needed to have MEWwalletKit conform to EIP1024. This is a first draft with bugs (the new unit test testEncode() crashes at two different places seemingly random (unknown object release and an out-of-bounds index).

Questions for the maintainers: 
- Is the current approach the right one?
- What is the best way to debug the code calling NaCl?

### CODE REVIEW

**Comment by Foboz** [comment]
Hey @ronaldmannak,

> Is the current approach the right one?

Everything should be fine, since it's aligned with decryption, except that I don't think that we need to have public visibility of key sizes

> crashes at two different places seemingly random

It's mostly because of wrong buffer sizes (encryption buffer should have 32 extra bytes, while encrypted data - 16). Let me double-check that and I'll comment what needs to be fixed

P.S. Also, FYI, we're moving repos to a new location https://github.com/mewwallet/, but it's ok to finish this PR here (since it's not "deprecated" yet) and I'll merge it into mewwallet one

**Comment by Foboz** [comment]
In addition to https://github.com/MyEtherWallet/mew-wallet-ios-kit/pull/7 and key sizes

I think it better to have helper function, like `generateNonce` inside of `twetnacl` package, and do not think about that in `mew-wallet-kit`

File: Sources/MEWwalletTweetNacl/TweetNacl.swift [hunk 1]
```diff
@@ -115,4 +124,27 @@ public class TweetNacl {
...
+    let k = try before(publicKey: theirPublicKey, secretKey: mySecretKey)
+    guard let data = message.data(using: .utf8) else { throw TweetNaclError.tweetNacl("Invalid message")}
+    return try secretbox(message: data, nonce: nonce, key: k)
+  }
+
+  private static func secretbox(message: Data, nonce: Data, key: Data) throws -> Data {
+    guard key.count == Constants.SecretBox.keyLength else { throw TweetNaclError.invalidKey }
+    guard nonce.count == Constants.SecretBox.nonceLength else { throw TweetNaclError.invalidNonce }
+      
+    var mData = Data(count: message.count + Constants.SecretBox.boxZeroLength)
```
Comment (Foboz) [inline Sources/MEWwalletTweetNacl/TweetNacl.swift:138]: should be `Constants.SecretBox.zeroLength`

File: Sources/MEWwalletTweetNacl/TweetNacl.swift [hunk 2]
```diff
@@ -115,4 +124,27 @@ public class TweetNacl {
...
+    mData.replaceSubrange(Constants.SecretBox.boxZeroLength..<mData.count, with: message)
```
Comment (Foboz) [inline Sources/MEWwalletTweetNacl/TweetNacl.swift:139]: also `Constants.SecretBox.zeroLength`

File: Sources/MEWwalletTweetNacl/TweetNacl.swift [hunk 3]
```diff
@@ -115,4 +124,27 @@ public class TweetNacl {
...
+    let m = [UInt8](mData)
+    var c = [UInt8](repeating: 0, count: message.count)
```
Comment (Foboz) [inline Sources/MEWwalletTweetNacl/TweetNacl.swift:141]: `mData.count`, not `message.count`

File: Sources/MEWwalletTweetNacl/TweetNacl.swift [hunk 4]
```diff
@@ -115,4 +124,27 @@ public class TweetNacl {
...
+    var nonce = [UInt8](nonce)
+    var key = [UInt8](key)
+      
+    let result = crypto_secretbox_xsalsa20poly1305_tweet(&c, m, UInt64(m.count), &nonce, &key)
+    guard result == 0 else { throw TweetNaclError.tweetNacl("[TweetNacl.secretbox] Internal error code: \(result)") }
+    
+    return Data(c[0..<Constants.SecretBox.zeroLength])
```
Comment (Foboz) [inline Sources/MEWwalletTweetNacl/TweetNacl.swift:148]: and `Data(c[Constants.SecretBox.boxZeroLength..<c.count])`

File: Sources/MEWwalletTweetNacl/TweetNacl.swift (hunk 2 above)
Comment (ronaldmannak) [reply Sources/MEWwalletTweetNacl/TweetNacl.swift:139]: Which one does this refer to? Line 139? 
Should it be `mData.replaceSubrange(0..<Constants.SecretBox.boxZeroLength, with: message)`?

File: Sources/MEWwalletTweetNacl/TweetNacl.swift (hunk 1 above)
Comment (ronaldmannak) [reply Sources/MEWwalletTweetNacl/TweetNacl.swift:138]: Changed to `key.count == Constants.SecretBox.zeroLength`

File: Sources/MEWwalletTweetNacl/TweetNacl.swift (hunk 2 above)
Comment (ronaldmannak) [reply Sources/MEWwalletTweetNacl/TweetNacl.swift:139]: Changed to `guard nonce.count == Constants.SecretBox.zeroLength...`

File: Sources/MEWwalletTweetNacl/TweetNacl.swift (hunk 3 above)
Comment (ronaldmannak) [reply Sources/MEWwalletTweetNacl/TweetNacl.swift:141]: This is the declaration of mData, so it doesn't have a length yet. Should we convert `message` to `utf8 Data` and use `messageData.count`?

**Comment by ronaldmannak** [comment]
Thanks @Foboz. I've updated the code. The unit test encrypts w/o error, but then throws an internal error code upon decryption. It looks like encryption is still not correct. Did I update everything correctly?

File: Sources/MEWwalletTweetNacl/TweetNacl.swift (hunk 2 above)
Comment (Foboz) [reply Sources/MEWwalletTweetNacl/TweetNacl.swift:139]: @ronaldmannak 
> `mData.replaceSubrange(Constants.SecretBox.boxZeroLength..<mData.count, with: message)`

this part still incorrect, should be

`mData.replaceSubrange(Constants.SecretBox.boxZeroLength..<mData.count, with: message)`

**Comment by ronaldmannak** [comment]
@Foboz Thanks for the second review, I appreciate it. The code now seems to work and is ready for review. I've added new unit tests as well. I will update the MEWwalletKit API next.

**Comment by Foboz** [comment]
Hi @ronaldmannak, thanks for the update. Looks great! However, do you think that we really need to have keys size in `public`? If so, I would prefer to update that part before merge

**Comment by ronaldmannak** [comment]
@Foboz My bad, I forgot to change those back. Done.

**Comment by Foboz** [comment]
Hey @ronaldmannak, sorry for being silent. I'm making last checks and will merge

**Pull request merged**